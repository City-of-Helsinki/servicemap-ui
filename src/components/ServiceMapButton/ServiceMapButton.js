import React from 'react';
import PropTypes from 'prop-types';
import { ButtonBase, Typography } from '@material-ui/core';
import { FormattedMessage } from 'react-intl';

// ServiceMapButton
const SMButton = (props) => {
  const {
    'aria-label': ariaLabel,
    children,
    classes,
    intl,
    className,
    small,
    color,
    icon,
    messageID,
    onClick,
    margin,
    style,
    role,
    ...rest
  } = props;
  const colorStyle = classes[color] || '';
  const buttonIcon = icon ? React.cloneElement(icon, { className: classes.buttonIcon }) : null;
  const buttonClasses = `${classes.button} ${small ? classes.smallButton : ''} ${margin ? classes.margin : classes.marginRight} ${className} ${colorStyle}`;
  const textClasses = classes.typography;

  let buttonTitle = null;

  if (messageID) {
    buttonTitle = intl.formatMessage({ id: messageID });
  }

  return (
    <ButtonBase
      {...rest}
      aria-label={ariaLabel || buttonTitle}
      className={buttonClasses}
      icon={buttonIcon}
      onClick={onClick}
      role={role || 'button'}
      style={{
        ...style,
      }}
      variant="contained"
    >
      {
        icon
      }
      {
        messageID
        && (
          <Typography aria-hidden color="inherit" component="p" variant="caption" className={textClasses}>
            <FormattedMessage id={messageID} />
          </Typography>
        )
      }
      {
        !messageID
        && children
      }
    </ButtonBase>
  );
};

SMButton.propTypes = {
  'aria-label': PropTypes.string,
  icon: PropTypes.objectOf(PropTypes.any),
  className: PropTypes.string,
  color: PropTypes.oneOf(['primary', 'secondary', 'default']),
  margin: PropTypes.bool,
  small: PropTypes.bool,
  messageID: PropTypes.string,
  onClick: PropTypes.func.isRequired,
  style: PropTypes.objectOf(PropTypes.any),
  classes: PropTypes.objectOf(PropTypes.any).isRequired,
  children: PropTypes.node,
  role: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  intl: PropTypes.objectOf(PropTypes.any).isRequired,
};

SMButton.defaultProps = {
  'aria-label': null,
  children: null,
  className: '',
  small: false,
  color: 'default',
  icon: null,
  margin: false,
  messageID: null,
  style: null,
  disabled: false,
};

export default SMButton;
