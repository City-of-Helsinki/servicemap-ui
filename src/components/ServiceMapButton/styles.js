export default theme => ({
  button: {
    height: 38,
    padding: '0 11px',
    boxSizing: 'border-box',
    borderRadius: 2,
  },
  smallButton: {
    height: 34,
    padding: '0 4px',
  },
  margin: {
    marginTop: theme.spacing.unitDouble,
    marginBottom: theme.spacing.unitDouble,
  },
  marginRight: {
    marginRight: theme.spacing.unitDouble,
  },
  primary: {
    color: theme.palette.primary.contrastText,
    backgroundColor: theme.palette.primary.main,
    '&:hover': {
      backgroundColor: theme.palette.primary.light,
    },
    '&:disabled': {
      backgroundColor: '#cccccc',
      color: 'rgba(0, 0, 0, 0.5)',
    },
  },
  secondary: {
    color: theme.palette.secondary.contrastText,
    backgroundColor: '#353638',
    border: `0.5px solid ${theme.palette.secondary.contrastText}`,
    '&:hover': {
      backgroundColor: theme.palette.secondary.light,
    },
    '&:disabled': {
      backgroundColor: theme.palette.secondary.light,
      color: 'rgba(0, 0, 0, 0.5)',
    },
  },
  default: {
    color: theme.palette.white.contrastText,
    backgroundColor: theme.palette.white.main,
    border: `0.5px solid ${theme.palette.white.contrastText}`,
    '&:hover': {
      backgroundColor: theme.palette.white.light,
    },
    '&:disabled': {
      // backgroundColor: theme.palette.white.dark,
      borderColor: theme.palette.white.dark,
      color: 'rgba(0, 0, 0, 0.5)',
    },
  },
  typography: {
    textTransform: 'none',
    margin: theme.spacing.unit,
    fontSize: 14,
  },
  disabled: {
    backgroundColor: 'rgba(0, 0, 0, 0.12)',
    color: 'rgba(0, 0, 0, 0.26)',
  },
});
