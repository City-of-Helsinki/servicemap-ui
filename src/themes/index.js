import { createMuiTheme } from '@material-ui/core';

const focusIndicator = {
  outline: '2px solid transparent',
  boxShadow: '0 0 0 2px rgb(255, 255, 255), 0 0 0 6px rgb(71, 131, 235), 0 0 0 8px rgb(255, 255, 255)',
};

const focusIndicatorDark = {
  outline: '2px solid transparent',
  boxShadow: '0 0 0 2px rgb(255, 255, 255), 0 0 0 6px rgb(0, 0, 0), 0 0 0 8px rgb(255, 255, 255)',
};

const props = {
  // Globally disable all default mui focus effects
  MuiButtonBase: {
    disableRipple: true,
  },
  MuiButton: {
    disableRipple: true,
    disableFocusRipple: true,
  },
  MuiTypography: {
    variant: 'body2',
  },
};

const overrides = theme => ({
  MuiButtonBase: {
    root: {
      // Default keyboard focus indicator for buttons
      '&$focusVisible': theme === 'dark' ? focusIndicatorDark : focusIndicator,
    },
  },
  MuiInputBase: {
    // Default keyboard focus indicator for input fields
    root: { '&$focused': theme === 'dark' ? focusIndicatorDark : focusIndicator },
  },
  MuiListItemText: {
    root: {
      marginTop: 0,
      marginBottom: 0,
    },
  },
  MuiSvgIcon: {
    root: {
      fontSize: 24,
    },
  },
  PrivateSwitchBase: {
    root: {
      padding: 12,
    },
  },
});

const breakpoints = {
  values: {
    xs: 0,
    sm: 700,
    md: 900,
    lg: 1280,
    xl: 1920,
  },
};

const spacing = 8;

const custom = {
  body2light: {
    color: 'rgba(0,0,0,0.7)',
  },
};

const zIndex = {
  behind: -1,
  forward: 50,
  sticky: 51,
  infront: 900,
};

const typography = {
  useNextVariants: true,
  fontSize: 16,
  // Use the system font instead of the default Roboto font.
  fontFamily: [
    'Lato',
  ].join(','),
  body1: {
    fontSize: '1.043rem',
    letterSpacing: '0.03125rem', // 0.5px
    lineHeight: '1.75rem',
  },
  body2: {
    fontSize: '0.913rem',
    letterSpacing: '0.015625rem', // 0.25px
    lineHeight: '1.25rem',
  },
  subtitle1: {
    fontSize: '1.03rem',
    fontWeight: 'bold',
    letterSpacing: '0.00937rem', // 0.15px
    lineHeight: '1.5rem',
  },
  caption: {
    fontSize: '0.7725rem', // 12.36px
    fontWeight: 'bold',
    letterSpacing: '0.025rem', // 0.4px
    lineHeight: '1rem',
    color: 'rgba(0,0,0,0.6)',
  },
  h1: {
    fontSize: '6.259rem',
    letterSpacing: '-0.09375em', // -1.5px
  },
  h2: {
    fontSize: '3.862rem',
    fontWeight: 'bold',
    letterSpacing: '-0.03125rem', // -0.5px
  },
  h3: {
    fontSize: '3.089rem',
    fontWeight: 'bold',
    letterSpacing: 0,
  },
  h4: {
    fontSize: '2.214rem',
    letterSpacing: '0.015625rem', // 0.25px
  },
  h5: {
    fontSize: '1.564rem',
    letterSpacing: 0,
  },
  h6: {
    fontSize: '1.288rem',
    fontWeight: 'bold',
    letterSpacing: '0.015625rem', // 0.25px
    lineHeight: '1.5rem',
  },
  button: {
    fontSize: '0.90125rem', // 14.42px
    fontWeight: 'bold',
    letterSpacing: '0.078125rem', // 1.25px
    lineHeight: '1rem', // 16px
    color: '#ffffff',
  },
  overline: {
    fontSize: '0.7725rem', // 12.36px
    fontWeight: 'bold',
    letterSpacing: '0.125rem', // 2px
    lineHeight: '1rem',
  },
};

// Color palette for normal theme
export const paletteDefault = {
  primary: {
    main: 'rgb(25, 100, 230)',
    highContrast: '#fff',
  },
  secondary: {
    main: 'rgb(43, 47, 57)',
    hover: '#1d39ad',
  },
  background: {
    main: 'linear-gradient(340.58deg, rgb(11, 123, 237) 0%, rgb(20, 108, 232) 67.04%, rgb(25, 100, 230) 100%)',
    plain: 'rgb(20, 108, 232)',
    front: 'linear-gradient(340.58deg, rgb(11, 123, 237) 0%, rgb(20, 108, 232) 67.04%, rgb(25, 100, 230) 100%)',
  },
  white: {
    light: '#f2f2f2',
    main: '#FFFFFF',
    dark: '#949494',
    contrastText: '#000',
  },
  detail: {
    main: 'rgb(25, 100, 230)',
    alpha: 'rgba(25,100,230,0.5)',
  },
  disabled: {
    main: 'rgb(239, 239, 239)',
    strong: '#898989',
  },
  focusBorder: {
    main: 'rgb(71, 131, 235)',
  },
  warning: {
    main: '#AF0021',
  },
};

// Color palette for dark theme
export const paletteDark = {
  primary: {
    main: '#353638',
    highContrast: '#fff',
  },
  secondary: {
    main: 'rgb(43, 47, 57)',
    hover: '#5b5d61',
  },
  background: {
    main: '#4A4A4C',
    plain: '#4A4A4C',
    front: 'linear-gradient(326.21deg, rgba(0, 0, 0, 0.79) 0%, rgba(71, 71, 71, 0.79) 100%)',
  },
  white: {
    light: '#f2f2f2',
    main: '#FFFFFF',
    dark: '#949494',
    contrastText: '#000',
  },
  detail: {
    main: '#585B63',
    alpha: 'rgba(88,91,99,0.5)',
  },
  disabled: {
    main: 'rgb(239, 239, 239)',
    strong: '#898989',
  },
  warning: {
    main: '#AF0021',
  },
  focusBorder: {
    main: '#000',
  },
};

// Themes
const SMTheme = createMuiTheme({
  props,
  overrides: overrides('default'),
  breakpoints,
  typography,
  spacing,
  custom,
  palette: paletteDefault,
  zIndex,
});

const SMThemeDark = createMuiTheme({
  props,
  overrides: overrides('dark'),
  breakpoints,
  typography,
  spacing,
  custom,
  palette: paletteDark,
  zIndex,
});


export default { SMTheme, SMThemeDark };
