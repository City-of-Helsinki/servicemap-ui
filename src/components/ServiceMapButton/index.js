import { withStyles } from '@material-ui/core';
import { injectIntl } from 'react-intl';
import SMButton from './ServiceMapButton';
import styles from './styles';

export default injectIntl(withStyles(styles)(SMButton));
