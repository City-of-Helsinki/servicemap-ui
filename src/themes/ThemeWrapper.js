import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { ThemeProvider } from '@mui/material/styles';
import themes from '.';

const mapStateToProps = (state) => {
  const { user } = state;
  return {
    theme: user.theme,
  };
};

// Component to handle theme changes
const ThemeHandler = ({ children, theme }) => (
// Get correct theme setting from store
  <ThemeProvider theme={theme === 'dark' ? themes.SMThemeDark : themes.SMTheme}>
    {children}
  </ThemeProvider>
);
ThemeHandler.propTypes = {
  children: PropTypes.node,
  theme: PropTypes.string.isRequired,
};
ThemeHandler.defaultProps = {
  children: null,
};

const ThemeWrapper = connect(
  mapStateToProps,
)(ThemeHandler);

export default ThemeWrapper;
