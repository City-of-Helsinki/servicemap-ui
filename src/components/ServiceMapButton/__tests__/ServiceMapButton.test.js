// Link.react.test.js
import React from 'react';
import { IntlProvider } from 'react-intl';
import { createMount } from '@material-ui/core/test-utils';
import { MuiThemeProvider } from '@material-ui/core';
import ServiceMapButton from '../index';
import themes from '../../../../themes';

// Mock props for intl provider
const intlMock = {
  locale: 'en',
  messages: {
    'button.text': 'Button text',
  },
};

// Generic required props for ServiceMapButton
const buttonMockProps = {
  onClick: () => {},
  role: 'button',
};

// eslint-disable-next-line react/prop-types
const Providers = ({ children }) => (
  <IntlProvider {...intlMock}>
    <MuiThemeProvider theme={themes.SMTheme}>
      {children}
    </MuiThemeProvider>
  </IntlProvider>
);

describe('<ServiceMapButton />', () => {
  // let render;
  let mount;

  beforeEach(() => {
    // render = createRender({ wrappingComponent: Providers });
    mount = createMount({ wrappingComponent: Providers });
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('should work', () => {
    const component = mount(
      <ServiceMapButton {...buttonMockProps}>Test</ServiceMapButton>,
    );
    expect(component).toMatchSnapshot();
  });

  it('does show text using children', () => {
    const component = mount(
      <ServiceMapButton {...buttonMockProps}>Test</ServiceMapButton>,
    );
    expect(component.text()).toEqual('Test');
  });

  it('does show text using intl message id', () => {
    const component = mount(
      <ServiceMapButton {...buttonMockProps} messageID="button.text" />,
    );
    expect(component.text()).toEqual(intlMock.messages['button.text']);
  });

  it('simulates click event', () => {
    const mockCallBack = jest.fn();
    const component = mount(
      <ServiceMapButton
        {...buttonMockProps}
        onClick={mockCallBack}
      />,
    );
    component.find('button').simulate('click');
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });

  it('does render default accessibility attributes correctly', () => {
    const component = mount(
      <ServiceMapButton
        {...buttonMockProps}
        messageID="button.text"
      />,
    );
    const buttonBase = component.find('ButtonBase');
    const p = component.find('p');
    // Expect aria-label to be same as given text
    expect(buttonBase.props()['aria-label']).toEqual(intlMock.messages['button.text']);
    // Expect visible text to be hidden from screen readers
    expect(p.props()['aria-hidden']).toEqual(true);
    // Expect visible text to be same as aria-label
    expect(p.text()).toEqual(buttonBase.props()['aria-label']);
  });

  it('does use given accessibility attributes correctly', () => {
    const component = mount(
      <ServiceMapButton
        {...buttonMockProps}
        aria-label="Testing label"
        role="link"
      />,
    );
    const buttonBase = component.find('ButtonBase');
    expect(buttonBase.props().role).toEqual('link');
    expect(buttonBase.props()['aria-label']).toEqual('Testing label');
  });
});
