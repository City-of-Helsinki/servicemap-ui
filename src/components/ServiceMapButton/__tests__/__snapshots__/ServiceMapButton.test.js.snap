// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ServiceMapButton /> should work 1`] = `
<injectIntl(WithStyles(SMButton))
  onClick={[Function]}
  role="button"
>
  <WithStyles(SMButton)
    intl={
      Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatDateToParts": [Function],
        "formatDisplayName": [Function],
        "formatList": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatNumberToParts": [Function],
        "formatPlural": [Function],
        "formatRelativeTime": [Function],
        "formatTime": [Function],
        "formatTimeToParts": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getDisplayNames": [Function],
          "getListFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralRules": [Function],
          "getRelativeTimeFormat": [Function],
        },
        "locale": "fi",
        "messages": Object {
          "button.text": "Button text",
        },
        "onError": [Function],
        "textComponent": Symbol(react.fragment),
        "timeZone": undefined,
      }
    }
    onClick={[Function]}
    role="button"
  >
    <SMButton
      aria-label={null}
      className=""
      classes={
        Object {
          "button": "SMButton-button-1",
          "default": "SMButton-default-7",
          "margin": "SMButton-margin-3",
          "marginRight": "SMButton-marginRight-4",
          "primary": "SMButton-primary-5",
          "primaryFocus": "SMButton-primaryFocus-9",
          "secondary": "SMButton-secondary-6",
          "smallButton": "SMButton-smallButton-2",
          "typography": "SMButton-typography-8",
        }
      }
      color="default"
      disabled={false}
      icon={null}
      intl={
        Object {
          "defaultFormats": Object {},
          "defaultLocale": "en",
          "formatDate": [Function],
          "formatDateToParts": [Function],
          "formatDisplayName": [Function],
          "formatList": [Function],
          "formatMessage": [Function],
          "formatNumber": [Function],
          "formatNumberToParts": [Function],
          "formatPlural": [Function],
          "formatRelativeTime": [Function],
          "formatTime": [Function],
          "formatTimeToParts": [Function],
          "formats": Object {},
          "formatters": Object {
            "getDateTimeFormat": [Function],
            "getDisplayNames": [Function],
            "getListFormat": [Function],
            "getMessageFormat": [Function],
            "getNumberFormat": [Function],
            "getPluralRules": [Function],
            "getRelativeTimeFormat": [Function],
          },
          "locale": "fi",
          "messages": Object {
            "button.text": "Button text",
          },
          "onError": [Function],
          "textComponent": Symbol(react.fragment),
          "timeZone": undefined,
        }
      }
      margin={false}
      messageID={null}
      onClick={[Function]}
      role="button"
      small={false}
      style={null}
    >
      <WithStyles(ButtonBase)
        aria-label={null}
        className="SMButton-button-1  SMButton-marginRight-4  SMButton-default-7"
        disabled={false}
        focusVisibleClassName={null}
        icon={null}
        onClick={[Function]}
        role="button"
        style={Object {}}
        variant="contained"
      >
        <ButtonBase
          aria-label={null}
          centerRipple={false}
          className="SMButton-button-1  SMButton-marginRight-4  SMButton-default-7"
          classes={
            Object {
              "disabled": "MuiButtonBase-disabled-11",
              "focusVisible": "MuiButtonBase-focusVisible-12",
              "root": "MuiButtonBase-root-10",
            }
          }
          component="button"
          disableRipple={true}
          disableTouchRipple={false}
          disabled={false}
          focusRipple={false}
          focusVisibleClassName={null}
          icon={null}
          onClick={[Function]}
          role="button"
          style={Object {}}
          tabIndex="0"
          type="button"
          variant="contained"
        >
          <button
            aria-label={null}
            className="MuiButtonBase-root-10 SMButton-button-1  SMButton-marginRight-4  SMButton-default-7"
            disabled={false}
            icon={null}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            style={Object {}}
            tabIndex="0"
            type="button"
            variant="contained"
          >
            Test
          </button>
        </ButtonBase>
      </WithStyles(ButtonBase)>
    </SMButton>
  </WithStyles(SMButton)>
</injectIntl(WithStyles(SMButton))>
`;
